backend/manager/modules/restapi/jaxrs/src/main/java/org/ovirt/engine/api/restapi/resource/BackendHostCpuPinningsResource.java

```diff
@@ -0,0 +1,112 @@
+package org.ovirt.engine.api.restapi.resource;
+
+import java.util.List;
+import java.util.Set;
+
+import org.ovirt.engine.api.model.HostCpuPinning;
+import org.ovirt.engine.api.model.HostCpuPinnings;
+import org.ovirt.engine.api.model.Vm;
+import org.ovirt.engine.api.model.Vms;
+import org.ovirt.engine.api.resource.HostCpuPinningsResource;
+import org.ovirt.engine.api.restapi.types.Mapper;
+import org.ovirt.engine.core.common.businessentities.VDS;
+import org.ovirt.engine.core.common.businessentities.VM;
+import org.ovirt.engine.core.common.businessentities.VdsCpuUnit;
+import org.ovirt.engine.core.common.queries.IdQueryParameters;
+import org.ovirt.engine.core.common.queries.QueryReturnValue;
+import org.ovirt.engine.core.common.queries.QueryType;
+import org.ovirt.engine.core.common.utils.CpuPinningHelper;
+import org.ovirt.engine.core.compat.Guid;
+
+public class BackendHostCpuPinningsResource extends AbstractBackendCollectionResource<HostCpuPinning, Object> implements HostCpuPinningsResource {
+
+    private Guid hostId;
+
+    public BackendHostCpuPinningsResource(Guid hostId) {
+        super(HostCpuPinning.class, Object.class);
+        this.hostId = hostId;
+    }
+
+    @Override
+    public HostCpuPinnings list() {
+        VDS host = getHost();
+        List<VM> vms = getMonitoredVms();
+
+        HostCpuPinnings pinnings = new HostCpuPinnings();
+        Mapper<VdsCpuUnit, HostCpuPinning> mapper = getMapper(VdsCpuUnit.class, HostCpuPinning.class);
+        for (VdsCpuUnit cpuUnit : host.getCpuTopology()) {
+            pinnings.getHostCpuPinnings().add(mapper.map(cpuUnit, null));
+        }
+
+        addPinnedVms(pinnings.getHostCpuPinnings(), vms);
+        addVdsAffinity(pinnings.getHostCpuPinnings(), host.getVdsmCpusAffinity());
+
+        pinnings.getHostCpuPinnings().stream().forEach(pinning -> addLinks(pinning));
+        return pinnings;
+    }
+
+    private void addPinnedVms(List<HostCpuPinning> hostPinnings, List<VM> vms) {
+        for (VM vm : vms) {
+            String vmPinning = CpuPinningHelper.getVmPinning(vm);
+            if (vmPinning == null) {
+                continue;
+            }
+            Set<Integer> vmPinnedCpus = CpuPinningHelper.getAllPinnedPCpus(vmPinning);
```

Comment:
`hostPinnings` should hold the VM IDs that uses it. Do you re-do the logic here to save synchronization on backend?

Reply:
Yes, that is one of the 2 reasons. The second reason is that by loading all monitored VMs at once, I can avoid calling the DB for every VM id to get the vm's name and cpu pinning policy (not yet set to rest vm object). And on the frontend, I just ask for the pinnings and I get also all the VMs names with one call.